### DISCUSSION TITLE
named workspaces as dynamics

### DISCUSSION THREAD

**Discussion by aserowy** [YaLTeR/niri#517] [Ideas]
Heho,

i switched to niri recently and it is a pleasure to work with it. I have one thing that bugs me. Coming from Hyprland/Sway my workflow resolved around creating and focusing workspaces by name.

Thus, creating workspaces by name with the given cli would be awesome. If this function would fit your vision, i can work/pr on the implementation with a small hint where i should start looking.

Kind regards
Alexander

**Comment by YaLTeR** [comment]
Hm, what's the workflow exactly?

**Reply by aserowy** [reply]
I would use rofi to query all workspaces and show their name. When ill need a new one ill type the name in rofi and enter. A script gets executed calling niri cli to create and focus the workspace by name. After that the workspace name gets shown and i can select it with rofi to focus it directly.

E.g. i need a new 'context' for a project. Ill call my workspace rofi, enter the new name 'bazinga' and enter to create it. After switching to other workspaces ill call my workspace rofi and type 'baz' to select it and enter to focus it.

This way i have direct context while selecting workspaces and can create these on the run.

**Reply by YaLTeR** [reply]
I see. I guess for now you can work around by editing your niri config with a script at runtime to add/remove workspace "something" {} lines.

**Reply by aserowy** [reply]
Would be a cool idea but immutable distro :D

**Reply by YaLTeR** [reply]
Even the niri config??

**Reply by aserowy** [reply]
@YaLTeR Would this function be of interest to you? I could work on an integration.

**Reply by YaLTeR** [reply]
I'm not sure how exactly it should work and integrate with the config. One idea is to make IPC changes transient, like `niri msg output`: if any named workspaces change in the config, then all IPC changes are dropped, but as long as you don't change named workspaces in the config, IPC changes persist.

Then, what should the actual commands be? `niri msg create-named-workspace "name"` (which creates one at the very top, since it's the same code that is used by the configured named workspaces) or `niri msg set-workspace-name "name"` (which sets the name of the currently focused workspace that you need to create beforehand)? Then, `niri msg unset-workspace-name` to unname?

cc @algernon what do you think?

**Reply by algernon** [reply]
I can see both `create-named-workspace` and `set-workspace-name` justified, for slightly different workflows. `set-workspace-name` would be handy when you already collected a number of windows on a dynamic workspace, and want to name it for easier navigation. `create-named-workspace` would be useful when you know in advance you want to collect windows in a (temporary) named workspace.

I can see myself using `set-workspace-name` =)

**Reply by aserowy** [reply]
Can you give me a rough overview where to start? Would love to get a little bit invested into working on niri.

**Reply by algernon** [reply]
My recommendation would be to start with adding the commands - and have them do nothing - first. If you `git grep` for `Action::`, that'll give you a lot of pointers. Then, you can implement `unset-workspace-name` relatively easily, because `layout.unname_workspace()` is already there.

Then, you will likely need an `add_named_workspace()` function in `src/layout/monitors.rs`. Searching for `self.workspaces.push` will give you a few hints how to accomplish that. With that function present in Monitors, implementing `create-named-workspace` should be relatively easy.

For naming a workspace: that's very similar to how `unname_workspace` is implemented: rather than setting `self.name` to `None` in `workspace.unname()`, it'd set it to `Some(string)`, and then build a few wrappers around it to be able to bind the operation to an IPC action.

Hope this helps!

**Comment by tv42** [comment]
I just want to point out that the Overview mode in Niri v25.05 makes creating workspaces "in the middle" of existing ones darn simple. It'd be great to be able to name them too.

I often create a new workspace to e.g. research a specific topic for a few hours.

**Comment by tv42** [comment]
I just discovered `niri msg action set-workspace-name FOO` and I think it's exactly what was needed! Thank you!
